runtime/lua/vim/filetype/detect.lua

```diff
@@ -1204,4 +1286,269 @@ end
 -- luacheck: pop
 -- luacheck: pop
 
+local patterns_hashbang = {
+  ['^zsh\\>'] = { 'zsh', { vim_regex = true } },
+  ['^\\(tclsh\\|wish\\|expectk\\|itclsh\\|itkwish\\)\\>'] = { 'tcl', { vim_regex = true } },
+  ['^expect\\>'] = { 'expect', { vim_regex = true } },
+  ['^gnuplot\\>'] = { 'gnuplot', { vim_regex = true } },
+  ['make\\>'] = { 'make', { vim_regex = true } },
+  ['^pike\\%(\\>\\|[0-9]\\)'] = { 'pike', { vim_regex = true } },
+  lua = 'lua',
+  perl = 'perl',
+  php = 'php',
+  python = 'python',
+  ['^groovy\\>'] = { 'groovy', { vim_regex = true } },
+  raku = 'raku',
+  ruby = 'ruby',
+  ['node\\(js\\)\\=\\>\\|js\\>'] = { 'javascript', { vim_regex = true } },
+  ['rhino\\>'] = { 'javascript', { vim_regex = true } },
+  -- BC calculator
+  ['^bc\\>'] = { 'bc', { vim_regex = true } },
+  ['sed\\>'] = { 'sed', { vim_regex = true } },
+  ocaml = 'ocaml',
+  -- Awk scripts; also finds "gawk"
+  ['awk\\>'] = { 'awk', { vim_regex = true } },
+  wml = 'wml',
+  scheme = 'scheme',
+  cfengine = 'cfengine',
+  escript = 'erlang',
+  haskell = 'haskell',
+  clojure = 'clojure',
+  ['scala\\>'] = { 'scala', { vim_regex = true } },
+  -- Free Pascal
+  ['instantfpc\\>'] = { 'pascal', { vim_regex = true } },
+  ['fennel\\>'] = { 'fennel', { vim_regex = true } },
+  -- MikroTik RouterOS script
+  ['rsc\\>'] = { 'routeros', { vim_regex = true } },
+  ['fish\\>'] = { 'fish', { vim_regex = true } },
+  ['gforth\\>'] = { 'forth', { vim_regex = true } },
+  ['icon\\>'] = { 'icon', { vim_regex = true } },
+}
+
+---@private
+-- File starts with "#!".
+local function match_from_hashbang(contents, path)
+  local first_line = contents[1]
+  -- Check for a line like "#!/usr/bin/env {options} bash".  Turn it into
+  -- "#!/usr/bin/bash" to make matching easier.
+  -- Recognize only a few {options} that are commonly used.
+  if matchregex(first_line, [[^#!\s*\S*\<env\s]]) then
+    first_line = first_line:gsub('%S+=%S+', '')
+    first_line = first_line
+      :gsub('%-[iS]', '', 1)
+      :gsub('%-%-ignore%-environment', '', 1)
+      :gsub('%-%-split%-string', '', 1)
```

Comment:
```suggestion
      :gsub('%-%-ignore%-environment', '', 1)
      :gsub('%-%-split%-string', '', 1)
      :gsub('%-[iS]', '', 1)
```
Starting with the most specific pattern fixes this particular instance, but might not be a general solution.

Reply:
But probably good for now (although a `TODO` comment might be good?)

Reply:
Good catch. Tbh, I am not sure why the Vim version works here: if the first pattern `-[iS]` matches, wouldn't the `-i` in  `--ignore-environment` be matched and consequently replaced first? So I think moving the less specific pattern `ignore-environment` to the end is definitely "more correct" and can probably stay that way. I can add a comment though.

Reply:
Same; I suspect it's some arcane codepath in `substitute` specifically, which may work "backwards" if multiple branches match?